Replace deprecated createCipher with createCipheriv for new tokens

gh2k opened the pull request:
This creates a new cipher label `2019.1` and uses it for all new encrypted tokens. These tokens have an additional 4th field which is the initialisation vector for the cipher.

Decryption is backwards-compatible and uses the old method for older labels.

Originally paired with @40thieves in tpds. Bug: overleaf/issues#2425

After merge, should be pulled into both tpds and third-party-references.

gh2k committed:
Run tests via 'npm test'

gh2k committed:
Replace deprecated createCipher with createCipheriv for new tokens

Co-authored-by: Alasdair Smith <ali@alasdairsmith.co.uk>

das7pad commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
+			salt = results[0]
+			iv = results[1]
+
+			keyFn cipherPassword, salt, 32, (err, key) =>
```
```suggestion
			keyFn @cipherPassword, salt, 32, (err, key) =>
```
bad reference

das7pad commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
 		return callback(new Error("invalid password")) if not password? or password.length < 16
-		keyFn password, new Buffer(salt, 'hex'), (err, key) =>
+
+		keyLength = if label == "2019.1" then 32 else 64
```
The conditional is error prone. For the next update, it would use a 64 char buffer for decryption but the encryption is hard coded at 32.

Using a second map in the settings for the `keyLength` could be an option. Moving the `keyLength` logic into the `keyFn` function is another option.

das7pad commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
 				return callback(err)
-			decipher = crypto.createDecipher(ALGORITHM, key)
+
+			decipher = if label == "2019.1"
```
Similar problem as above.

das7pad commented on package-lock.json:
```diff
@@ -673,74 +1414,37 @@
-      }
+      "dev": true
     },
-    "mocha": {
```
Mocha is a peerDependency of `grunt-mocha-test` now.

In a clean checkout of this branch I ran `npm ci`, but mocha was missing - so a quick `npm i mocha` followed (and another `npm i` to get sinon).
While we are fixing dependencies, maybe include these two as well? `npm i --save-dev mocha` and `npm i --save-dev sinon@...` any version will likely work.

gh2k committed:
Update lib/coffee/AccessTokenEncryptor.coffee

Co-Authored-By: Jakob Ackermann <das7pad@outlook.com>

gh2k committed:
Update dev dependencies

gh2k commented:
Thanks for your comments. Good spot on the invalid reference. :+1: 

I'm leaving the conditions as-is for now. `createCipheriv` fails if the key length is incorrect, where `createCipher` didn't (which is why it accepted a key that was too long) so we'd notice pretty quickly if we added a new label and forgot to make the change here. I can't see us doing this on a regular basis.

briangough reviewed (changes requested):
I don't think this is quite right, in `decryptToJson` we should not use the cipher label to determine the keylength, but the absence of the `iv` field.  If there are 3 fields (old-style) do the decryption the old way, if there are 4 fields (new-style) do it the new way.

To make it easy to audit the code, if possible retain the main body of the existing decryptToJson code (e.g. as `decryptToJsonV1`)  and after checking the number of fields either call decryptToJsonV1 or a new version decryptToJsonV2

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
+		async.parallel [
+			(cb) -> crypto.randomBytes(16, cb)
+			(cb) -> crypto.randomBytes(16, cb)
+		], (err, results) =>
```
I think you may as well skip the `async.parallel` here and do `crypto.randomBytes(32, function (err, buff) { salt = buf.slice(0,8) iv = buf.slice(8,16) ...}`. It's a nice improvement to use the asynchronous crypto function though.

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
-keyFn = (password, salt, callback)->
-	return crypto.pbkdf2(password, salt, 10000, 64, 'sha1', callback)
+keyFn = (password, salt, keyLength, callback)->
+	return crypto.pbkdf2(password, salt, 10000, keyLength, 'sha1', callback)
```
Let's keep the original `keyFn` and add a new one.

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
 		@settings = settings
-		@cipherLabel = @settings.cipherLabel
-		throw Error("cipherLabel must not contain a colon (:)") if @cipherLabel?.match(/:/)
+		@cipherLabel = "2019.1"
```
We should allow this to come through settings, so each application can have its own current cipherLabel.

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,42 +1,61 @@
+				crypto.createDecipheriv(ALGORITHM, key, Buffer.from(iv, 'hex'))
+			else
+				crypto.createDecipher(ALGORITHM, key)
+
```
If we successfully decrypt the token, and it's an old version, we should upgrade it to the new one.

gh2k replied:
I'm not sure how we would do this? We aren't aware of the caller's implementation at this point, and the calling method is only expecting a deciphered object. It should I think be up to the caller to do this as it knows where to store the updated object.

briangough replied:
You are right, it has to be done in the caller, we can signal it here by returning an extra parameter `upgradeNeeded = (label isnt @cipherLabel)`

jdleesmiller commented:
> Good spot on the invalid reference.

While I don't want to create more work... should we start by decaf-ing this so we can use a (better) linter?

(And yes, good spot!)

gh2k committed:
Keep oldmethods and use old method when app uses an older label

gh2k commented:
I think I've addressed the comments here. I've re-added the old methods and called into the new methods where appropriate. Also used the field count to decide whether to use the new methods and kept the behaviour the same if the caller uses an older label.

> decaf

I'd rather do this separately so as not to confuse things too much, but agree that it needs doing.

das7pad commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -1,10 +1,14 @@
 crypto = require('crypto')
+async = require('async')
```
```suggestion
```

It was missing in the `dependencies` previously, so there is no need for another deps update.

gh2k replied:
:+1: thanks

gh2k committed:
Remove unneeded 'async' dependency

Co-Authored-By: Jakob Ackermann <das7pad@outlook.com>

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -18,6 +22,9 @@ class AccessTokenEncryptor
 		throw Error("cipherPassword too short") if @cipherPassword.length < 16
 
 	encryptJson: (json, callback) ->
+		unless ["2015.1", "2016.1"].includes(@cipherLabel)
```
I don't think this needs to be conditional.  We can always write encrypt the password in the new format (because the format and the label are orthogonal).  Even if we have an old label in use we can still store the encrypted result in the new format - so we only need to encrypt in the new format.

gh2k replied:
Right, yes I've been thinking about the label incorrectly. This makes sense.

briangough commented on lib/coffee/AccessTokenEncryptor.coffee:
```diff
@@ -28,11 +35,31 @@ class AccessTokenEncryptor
+		[label, salt, cipherText, iv] = encryptedJson.split(':', 4)
+		if iv and iv.length > 0
+			return @decryptToJsonV2(encryptedJson, callback)
+
```
I feel like this would be clearer if we used

```coffee
[label, salt, cipherText, iv] = encryptedJson.split(':', 4)
password = @settings.cipherPasswords[label]
return callback(new Error("invalid password")) if not password? or password.length < 16
if iv
  ... decrypTtoJsonV2(password, label, salt, cipherText, iv)
else
  ... decryptToJsonV1(password, label, salt, ciphertext)
```

to avoid duplication of splitting the string and checking the password.
No need to check iv.length here as it is a string, so any non-empty string will evaluate to `true`.

gh2k replied:
>  No need to c No need to check iv.length here as it is a string, so any non-empty string will evaluate to true.heck iv.length here as it is a string, so any non-empty string will evaluate to true.

Showing my `c` roots again, aren't I?

briangough reviewed (changes requested):
Thanks for the changes, just a few tweaks neeeded to avoid unnecessary code.

gh2k committed:
Split methods to reduce code duplication

briangough reviewed (changes requested):
Can we have a unit test for the decryption of the new format? I think we aren't testing decryption with an `iv` field.

gh2k commented:
> Can we have a unit test for the decryption of the new format? I think we aren't testing decryption with an `iv` field.

This is covered here:
https://github.com/overleaf/access-token-encryptor/blob/78620f59ba76189b42e04ea4b088fba62d651f8e/test/unit/coffee/AccessTokenEncryptorTests.coffee#L41-L47

We know that we're supplying an `iv` field for this as we've tested for that above.

gh2k committed:
Add an explicit test for a 2019 string
